### ISSUE TITLE
Adding Julia extension instructions to template

### ISSUE THREAD

**Issue by snairdesai** [gentzkow/template#76]
The purpose of this issue (#76) is to provide additional user documentation in `README.md` on our Julia extension, which was integrated within `gentzkow/template` in #67. I noticed when pulling these changes to our `selective-exposure` project that the `README.md` file does not currently have instructions on running Julia scripts or initializing the Julia environment within `miniconda`. This process differs slightly from those we use for other softwares, because Julia is not yet fully integrated within `conda-forge`. I will shortly post a proposed revision to the `README.md` file describing this procedure. 

cc @jc-cisneros @gentzkow

**Comment by snairdesai** [comment]
I propose we edit the following sections as follows (new additions are in bold):

----

### [Requirements](https://github.com/gentzkow/template/blob/master/README.md#requirements)

**_Note:_** The application requirements and setup instructions outlined below are intended to serve general users. To build the repository as-is, the following applications are required:

* [R](https://cran.r-project.org/mirrors.html)
* [Stata](https://www.stata.com/install-guide/)
* [Python](https://www.python.org/downloads/)
* [git](https://git-scm.com/download/mac)
* [git lfs](https://git-lfs.github.com/)
* [LyX](https://www.lyx.org/Download)
* A TeX distribution for your local OS (for example, [MacTeX](https://www.tug.org/mactex/) for MacOS).

You may download the latest versions of each. By default, the **[Setup](#setup)** instructions below will assume their usage. Note that some of these applications must also be invocable from the command line. See the **[Command Line Usage](#command-line-usage)** section for details on how to set this up. **Note that if you wish to run `Julia` scripts in your repository, you will additionally need to [install `Julia`](https://julialang.org/downloads/) and set up its command line usage. Julia is currently not required to build the repository as-is.**

You must set up a personal `GitHub` account to [clone private repositories](https://docs.github.com/en/account-and-profile/setting-up-and-managing-your-personal-account-on-github/managing-access-to-your-personal-repositories) on which you are a collaborator. For public repositories (such as `template`), `Git` will suffice. You may need to set up [Homebrew](https://brew.sh/) if `git` and `git-lfs` are not available on your local computer.

If you are using MacOS, [ensure your terminal is operating in bash](https://www.howtogeek.com/444596/how-to-change-the-default-shell-to-bash-in-macos-catalina/) rather than the default `zsh`. MacOS users who are running `template` on an **Apple Silicon chip** will instead want to use `Rosetta` as their default terminal. You can find instructions on how to shift from `zsh` to `Rosetta` [here](https://osxdaily.com/2020/12/04/how-install-rosetta-2-apple-silicon-mac/) and [here](https://www.courier.com/blog/tips-and-tricks-to-setup-your-apple-m1-for-development/). WindowsOS users (with Version 10 or higher) will need to switch to `bash` from `PowerShell`. To do this, you can run `bash` from within a `PowerShell` terminal (you must have installed `git` first).

Once you have met these OS and application requirements, [clone a team repository](https://docs.github.com/en/repositories/creating-and-managing-repositories/cloning-a-repository) from `GitHub` and proceed to **[Setup](#setup)**.

----

### [Adding Packages](https://github.com/gentzkow/template/blob/master/README.md#adding-packages)

_Note_: These instructions are relevant for users who have installed `miniconda`. If you have not done so, consult `/setup/dependencies.md`.

#### _Python_
Add any required packages to `/setup/conda_env.yaml`. If possible add the package version number as well. If there is a package that is not available from `conda`, add this to the `pip` section of the `yaml` file. In order to not re-run the entire environment setup you can download these individual files from `conda` with the command:

```
conda install -c conda-forge --name <environment name> <package_name=version_number>
```
#### _R_
Add any required packages that are available via `CRAN` to `/setup/conda_env.yaml`. These must be prepended with `r-`. If there is a package that is only available from `GitHub` and not from `CRAN`, add this package to `/setup/setup_r.r` (after copying this script from `/extensions`). These individual packages can be added in the same way as `Python` packages above (with the `r-` prepend).

#### _Stata_

Install `Stata` dependencies using `/setup/download_stata_ado.do` (copy `download_stata_ado.do` from `/extensions` to `/setup` first). We keep all non-base `Stata` ado files in the `lib` subdirectory, so most non-base `Stata` ado files will be versioned. To add additional `Stata` dependencies, use the following `bash` command from the `setup` subdirectory:

```
stata-mp -e download_stata_ado.do
```

#### _Julia_

**First, add any required Julia packages to `julia_conda_env.jl`. Note that these packages must be available via `conda-forge`. Follow the same steps described in **[Setup](#setup)** to build and activate your `conda` environment, being sure to *uncomment the line referencing `julia` in `/setup/conda_env.yaml`* before building the environment. Once the environment is activated, run the following in terminal:**

```
julia julia_conda_env.jl
```

----


### [Command Line Usage](https://github.com/gentzkow/template/blob/master/README.md#command-line-usage)

For instructions on how to set up command line usage, refer to the [repo wiki](https://github.com/gentzkow/template/wiki/Command-Line-Usage).

By default, the repository assumes these executable names for the following applications:

```
application : executable

python      : python
git-lfs     : git-lfs
lyx         : lyx
r           : Rscript
stata       : stata-mp (this will need to be updated if using a version of Stata that is not Stata-MP)
julia       : julia
```

Default executable names can be updated in `config_user.yaml`. For further details, see the **[User Configuration](#user-configuration)** section.

----

### [Windows Differences](https://github.com/gentzkow/template/blob/master/README.md#windows-differences)

The only addition here should be updating the WindowsOS executable name for `julia`, but we should check this by running on @jc-cisneros's Windows.

**Comment by snairdesai** [comment]
Additional flags for @jc-cisneros and I to complete:

- [ ] Add Julia as an additional executable for [`/setup/config_user_template.yaml`](https://github.com/gentzkow/template/blob/master/setup/config_user_template.yaml).
- [ ] Ensure the Julia extension works in the absence of a `conda` environment. This will likely require minor edits to `~/setup/dependencies.md` (i.e., to provide instructions for manual installations of Julia packages - in this case, simply by running `julia julia_conda_env.jl` in terminal) and `~/setup/check_setup.py` (i.e., to ensure CLI for Julia is enabled for users without an active `conda` environments).
- [ ] Revise the [Command Line Interface section of the Wiki](https://github.com/gentzkow/template/wiki/Command-Line-Usage) to describe Julia CLI setup for those not using `conda`.
- [ ] Ensure the Julia setup works on WindowsOS.
- [ ] In a separate issue, we may consider reproducing `template` to run directly using `Julia` rather than `python` (i.e., [adding an `extensions\julia` subfolder](https://github.com/gentzkow/template/tree/master/extensions)similar to what currently exists for `STATA` and `R`. If we choose to do this at some point, we would also need to revise [Running Package Scripts in Other Languages](https://github.com/gentzkow/template/blob/master/README.md#running-package-scripts-in-other-languages) in the `README.md`.

We can either accomplish the above here, or create a new issue.

cc @jc-cisneros

**Comment by jc-cisneros** [comment]
@snairdesai thanks for this! The additions make sense. The only one I would change is `Note that these packages must be available via conda-forge`. My understanding if the user chooses to use miniconda, then we are using two package managers in the following way:
- When the conda environment is built, it will contain an installation of Julia with no additional packages.
- When running `julia julia_conda_env.jl`, we are using [Julia Pkg package manager](https://juliapackages.com/). These packages get installed in the Julia directory connected to the Julia installation the user called in the command. In other words, if the conda environment is activated, these packages will get installed in the conda directory and not affect a local Julia installation.

The reason we are keeping track of the integration of Julia packages with Conda is that in the process above we are solving dependencies separately for Python/R packages and Julia packages. This will not solve dependencies or harmonize the paths across Python/R/Julia (this gave us [issues when using a matplotlib backend previously](https://github.com/gentzkow/template/pull/68#issuecomment-1285951640)).

**Comment by snairdesai** [comment]
Thanks @jc-cisneros! Got it -- makes sense. I'll remove this text from the `README.md`. Any preference on whether we should tackle the [above items](https://github.com/gentzkow/template/issues/76#issuecomment-1433611367)directly in this issue, or move them to elsewhere to handle at a later time?

**Comment by jc-cisneros** [comment]
@snairdesai Regarding the steps outlined in https://github.com/gentzkow/template/issues/76#issuecomment-1433611367, I would split them into separate issues that are handled in parallel to this one. In particular, I would suggest one issue dealing with using `template` with the user's own version of Julia, another one to report the results from any test with Windows OS, and a last one (if considered necessary) adding a version of template that is run with Julia. All of these parallel issues would feed input to make changes on the instructions, which is the scope of this issue.

**Comment by snairdesai** [comment]
### Summary + Deliverables

In this issue (#76), we updated the `README` to include additional instructions for `Julia` users (see [`48040d7`](https://github.com/gentzkow/template/commit/48040d7094ae045847b04b8760d8d19a3a5a8814) or [here](https://github.com/gentzkow/template/issues/76#issuecomment-1433611052) for a formatted version, and added instructions on manual package installations for non-conda users (see [`4cfb3eb`](https://github.com/gentzkow/template/commit/4cfb3ebd33fc1c575b0d2cf4e2ccae3a700c8a56)).

Following a discussion with @jc-cisneros, we determined that the additional action items referenced [above](https://github.com/gentzkow/template/issues/76#issuecomment-1433611367) should be addressed in separate issues, which we will open and resolve as more bandwidth arises. I will proceed to open a PR with @jc-cisneros as a reviewer.
